firestore-genai-chatbot/functions/src/generative-client/google_ai.ts

```diff
@@ -24,103 +25,57 @@ type ApiMessage = {
   }[];
 };
 
-enum Role {
-  USER = 'user',
-  GEMINI = 'model',
-}
+const ai = genkit({
+  plugins: [googleAI()],
+});
 
 export class GeminiDiscussionClient extends DiscussionClient<
-  GoogleGenerativeAI,
+  Genkit,
   GeminiChatOptions,
   ApiMessage
 > {
   modelName: string;
+
   constructor({apiKey, modelName}: {apiKey?: string; modelName: string}) {
-    super();
+    super(ai);
     if (!apiKey) {
       throw new Error('API key required.');
     }
     if (!modelName) {
       throw new Error('Model name required.');
     }
     this.modelName = modelName;
-    this.client = new GoogleGenerativeAI(apiKey);
-  }
-
-  createApiMessage(
-    messageContent: string,
-    role: 'user' | 'model' = 'user'
-  ): ApiMessage {
-    const apiRole = role === 'user' ? Role.USER : Role.GEMINI;
-
-    return {
-      role: apiRole,
-      parts: [{text: messageContent}],
-    };
   }
 
   async generateResponse(
-    history: Message[],
+    history: MessageData[],
     latestApiMessage: ApiMessage,
     options: GeminiChatOptions
   ) {
-    if (!this.client) {
-      throw new Error('Client not initialized.');
-    }
-
-    const model = this.client.getGenerativeModel({
-      model: this.modelName,
-    });
-
-    const chatSession = model.startChat({
-      history: this.messagesToApi(history),
-      generationConfig: {
-        topP: options.topP,
-        topK: options.topK,
-        temperature: options.temperature,
-        maxOutputTokens: options.maxOutputTokens,
-        candidateCount: options.candidateCount,
-      },
-      safetySettings: options.safetySettings,
-    });
-
-    let result;
     try {
-      result = await chatSession.sendMessage(latestApiMessage.parts[0].text);
+      const llmResponse = await this.client!.generate({
+        prompt: latestApiMessage.parts,
+        messages: history,
+        model: this.modelName,
+        config: {
+          topP: options.topP,
+          topK: options.topK,
+          temperature: options.temperature,
+          maxOutputTokens: options.maxOutputTokens,
+        },
+      });
+
+      return {
+        response: llmResponse.text,
+        // TODO how to handle candidates through genkit?
+        candidates: [],
+        // TODO might be in "evaluations API of genkit" or might be here
+        // safetyMetadata: llmResponse.
```

Comment:
I believe you'll need to extract the safety settings from the response somehow :/

I can add this to vertex if it's needed